File: musculo-skeletal/python/vitruvius/parser.py
"""Recursive-descent parser for Vitruvius.

The grammar is LL(1) (specification Proposition 3.2), so one token of
lookahead suffices everywhere and no backtracking is used.

Two syntactic facts are load-bearing and enforced here:

  * `outbound` and `return` are mandatory in a circuit declaration, so an
    open circuit cannot arise by omission -- only through an explicit
    `without` / `reroute` operator inside an experiment.
  * `excluding`-style optionality is absent: postfix lesion operators are
    parsed as a left-associative chain, matching the left-recursion
    elimination given in the specification.
"""

from __future__ import annotations

from .ast_nodes import (
    AntagonistDecl, CircuitDecl, CircuitRef, CompartmentDecl, ElementDecl,
    EventTypeDecl, ExperimentDecl, FloorSpec, InstanceDecl, LesionDecl,
    Observable, PhaseDecl, Program, Quantity, Reroute, TemplateDecl,
    WithNoise, WithScaling, WithoutElement, WithoutReturn,
)
from .lexer import STRATA, Tok, Token, tokenize


class ParseError(Exception):
    def __init__(self, msg: str, tok: Token):
        super().__init__(f"{msg} (got {tok.text!r} at line {tok.line}, col {tok.col})")
        self.token = tok


class Parser:
    def __init__(self, toks: list[Token]):
        self.toks = toks
        self.i = 0

    # ── token helpers ────────────────────────────────────────────────

    @property
    def cur(self) -> Token:
        return self.toks[self.i]

    def at(self, kind: Tok, text: str | None = None) -> bool:
        t = self.cur
        return t.kind == kind and (text is None or t.text == text)

    def at_kw(self, *words: str) -> bool:
        return self.cur.kind == Tok.KEYWORD and self.cur.text in words

    def eat(self, kind: Tok, text: str | None = None) -> Token:
        if not self.at(kind, text):
            want = text or kind.name
            raise ParseError(f"expected {want}", self.cur)
        t = self.cur
        self.i += 1
        return t

    def kw(self, word: str) -> Token:
        return self.eat(Tok.KEYWORD, word)

    def punct(self, ch: str) -> Token:
        return self.eat(Tok.PUNCT, ch)

    def ident(self) -> str:
        # Stratum names are keywords but are legal in identifier position
        # inside paths and argument lists.
        if self.cur.kind == Tok.KEYWORD and self.cur.text in STRATA:
            t = self.cur
            self.i += 1
            return t.text
        return self.eat(Tok.IDENT).text

    def quantity(self) -> Quantity:
        t = self.cur
        if t.kind == Tok.QUANTITY:
            self.i += 1
            return Quantity(float(t.value), t.unit)
        if t.kind == Tok.NUMBER:
            self.i += 1
            return Quantity(float(t.value), "")
        raise ParseError("expected a quantity", t)

    def number(self) -> float:
        t = self.cur
        if t.kind in (Tok.NUMBER, Tok.QUANTITY):
            self.i += 1
            return float(t.value)
        raise ParseError("expected a number", t)

    # ── program ──────────────────────────────────────────────────────

    def parse_program(self) -> Program:
        prog = Program()

        if self.at_kw("module"):
            self.kw("module")
            prog.module = self.ident()
            self.punct(";")

        while self.at_kw("import"):
            self.kw("import")
            prog.imports.append(self.ident())
            self.punct(";")

        while not self.at(Tok.EOF):
            if self.at_kw("compartment"):
                prog.compartments.append(self.parse_compartment())
            elif self.at_kw("circuit"):
                decl = self.parse_circuit_or_template_or_instance()
                if isinstance(decl, TemplateDecl):
                    prog.templates.append(decl)
                elif isinstance(decl, InstanceDecl):
                    prog.instances.append(decl)
                else:
                    prog.circuits.append(decl)
            elif self.at_kw("event"):
                prog.event_types.append(self.parse_event_type())
            elif self.at_kw("antagonist"):
                prog.antagonists.append(self.parse_antagonist())
            elif self.at_kw("experiment"):
                prog.experiments.append(self.parse_experiment())
            else:
                raise ParseError("expected a declaration", self.cur)

        return prog

    # ── compartment ──────────────────────────────────────────────────

    def parse_compartment(self) -> CompartmentDecl:
        ln = self.kw("compartment").line
        name = self.ident()
        self.punct("{")
        self.kw("capacitance")
        self.punct(":")
        cap = self.quantity()
        self.punct(";")
        self.kw("stratum")
        self.punct(":")
        if not self.at_kw(*STRATA):
            raise ParseError("expected a stratum name", self.cur)
        stratum = self.cur.text
        self.i += 1
        self.punct(";")
        self.punct("}")
        return CompartmentDecl(name, cap, stratum, ln)

    # ── circuit / template / instance ────────────────────────────────

    def parse_circuit_or_template_or_instance(self):
        ln = self.kw("circuit").line

        if self.at_kw("template"):
            self.kw("template")
            name = self.ident()
            self.punct("(")
            params: list[str] = []
            if not self.at(Tok.PUNCT, ")"):
                params.append(self.ident())
                while self.at(Tok.PUNCT, ","):
                    self.punct(",")
                    params.append(self.ident())
            self.punct(")")
            floor, outb, ret, elems = self.parse_circuit_body()
            return TemplateDecl(name, params, floor, outb, ret, elems, ln)

        name = self.ident()

        if self.at(Tok.PUNCT, "="):  # instance
            self.punct("=")
            tmpl = self.ident()
            self.punct("(")
            args: list[object] = []
            if not self.at(Tok.PUNCT, ")"):
                args.append(self.template_arg())
                while self.at(Tok.PUNCT, ","):
                    self.punct(",")
                    args.append(self.template_arg())
            self.punct(")")
            self.punct(";")
            return InstanceDecl(name, tmpl, args, ln)

        floor, outb, ret, elems = self.parse_circuit_body()
        return CircuitDecl(name, floor, outb, ret, elems, ln)

    def template_arg(self) -> object:
        if self.cur.kind in (Tok.NUMBER, Tok.QUANTITY):
            return self.quantity()
        return self.ident()

    def parse_circuit_body(self):
        self.punct("{")
        self.kw("floor")
        self.punct(":")
        floor = self.parse_floor_spec()
        self.punct(";")
        self.kw("outbound")
        self.punct(":")
        outb = self.parse_path()
        self.punct(";")
        self.kw("return")
        self.punct(":")
        ret = self.parse_path()
        self.punct(";")

        elems: list[ElementDecl] = []
        while self.at_kw("element"):
            elems.append(self.parse_element())
        self.punct("}")
        return floor, outb, ret, elems

    def parse_floor_spec(self) -> FloorSpec:
        if self.at_kw("derived"):
            self.kw("derived")
            self.punct("(")
            call = self.ident()
            arg = None
            if self.at(Tok.PUNCT, "("):
                self.punct("(")
                arg = self.ident()
                self.punct(")")
            self.punct(")")
            return FloorSpec(derived_call=call, derived_arg=arg)
        return FloorSpec(literal=self.quantity())

    def parse_path(self) -> list[str]:
        path = [self.ident()]
        while self.at(Tok.ARROW):
            self.eat(Tok.ARROW)
            path.append(self.ident())
        return path

    def parse_element(self) -> ElementDecl:
        ln = self.kw("element").line
        name = self.ident()
        self.kw("conducts")
        src = self.ident()
        self.eat(Tok.ARROW)
        dst = self.ident()

        delay = None
        gain = 1.0
        if self.at_kw("delay"):
            self.kw("delay")
            delay = self.quantity()
        if self.at_kw("gain"):
            self.kw("gain")
            gain = self.number()
        self.punct(";")
        return ElementDecl(name, src, dst, delay, gain, ln)

    # ── event types / antagonists ────────────────────────────────────

    def parse_event_type(self) -> EventTypeDecl:
        ln = self.kw("event").line
        self.kw("type")
        name = self.ident()
        self.punct("=")
        ctor = self.ident()
        args: list[str] = []
        if self.at(Tok.PUNCT, "("):
            self.punct("(")
            if not self.at(Tok.PUNCT, ")"):
                args.append(self.ident())
                while self.at(Tok.PUNCT, ","):
                    self.punct(",")
                    args.append(self.ident())
            self.punct(")")
        self.punct(";")
        return EventTypeDecl(name, ctor, args, ln)

    def parse_antagonist(self) -> AntagonistDecl:
        ln = self.kw("antagonist").line
        name = self.ident()
        self.punct("{")
        self.kw("agonist")
        self.punct(":")
        ago = self.ident()
        self.punct(";")
        self.kw("antagonist")
        self.punct(":")
        ant = self.ident()
        self.punct(";")
        self.kw("shared")
        self.punct(":")
        shared = [self.ident()]
        while self.at(Tok.PUNCT, ","):
            self.punct(",")
            shared.append(self.ident())
        self.punct(";")
        self.punct("}")
        return AntagonistDecl(name, ago, ant, shared, ln)

    # ── experiments ──────────────────────────────────────────────────

    def parse_experiment(self) -> ExperimentDecl:
        ln = self.kw("experiment").line
        name = self.ident()
        self.punct("{")
        self.kw("intact")
        self.punct(":")
        intact = self.parse_circuit_expr()
        self.punct(";")

        lesions: list[LesionDecl] = []
        observables: list[Observable] = []
        phases: list[PhaseDecl] = []

        while not self.at(Tok.PUNCT, "}"):
            if self.at_kw("lesion"):
                lesions.append(self.parse_lesion())
            elif self.at_kw("phase"):
                phases.append(self.parse_phase())
            elif self.at_kw("observe"):
                self.kw("observe")
                self.punct(":")
                observables = self.parse_observables()
                self.punct(";")
            else:
                raise ParseError("expected lesion, phase, or observe", self.cur)

        self.punct("}")

        if phases and (lesions or observables):
            raise ParseError(
                "an experiment is either phased or flat, not both", self.cur
            )
        return ExperimentDecl(name, intact, lesions, observables, phases, ln)

    def parse_phase(self) -> PhaseDecl:
        ln = self.kw("phase").line
        name = self.ident()
        from_phase = None
        if self.at_kw("from"):
            self.kw("from")
            from_phase = self.ident()
        self.punct("{")

        lesions: list[LesionDecl] = []
        observables: list[Observable] = []
        while not self.at(Tok.PUNCT, "}"):
            if self.at_kw("lesion"):
                lesions.append(self.parse_lesion())
            elif self.at_kw("observe"):
                self.kw("observe")
                self.punct(":")
                observables = self.parse_observables()
                self.punct(";")
            else:
                raise ParseError("expected lesion or observe", self.cur)
        self.punct("}")
        return PhaseDecl(name, lesions, observables, from_phase, ln)

    def parse_lesion(self) -> LesionDecl:
        ln = self.kw("lesion").line
        name = self.ident()
        self.punct(":")
        expr = self.parse_circuit_expr()
        self.punct(";")
        return LesionDecl(name, expr, ln)

    def parse_observables(self) -> list[Observable]:
        obs = [self.parse_observable()]
        while self.at(Tok.PUNCT, ","):
            self.punct(",")
            obs.append(self.parse_observable())
        return obs

    def parse_observable(self) -> Observable:
        ln = self.cur.line
        name = self.ident()
        args: list[str] = []
        if self.at(Tok.PUNCT, "("):
            self.punct("(")
            if not self.at(Tok.PUNCT, ")"):
                args.append(self.ident())
                while self.at(Tok.PUNCT, ","):
                    self.punct(",")
                    args.append(self.ident())
            self.punct(")")
        return Observable(name, args, ln)

    # ── circuit expressions ──────────────────────────────────────────

    def parse_circuit_expr(self):
        """Left-associative chain of postfix lesion operators."""
        ln = self.cur.line
        expr = CircuitRef(self.ident(), ln)

        while self.at_kw("without", "with", "reroute"):
            if self.at_kw("without"):
                self.kw("without")
                if self.at_kw("return"):
                    self.kw("return")
                    self.punct("(")
                    frm = self.ident()
                    self.punct(")")
                    expr = WithoutReturn(expr, frm, ln)
                elif self.at_kw("element"):
                    self.kw("element")
                    self.punct("(")
                    el = self.ident()
                    self.punct(")")
                    expr = WithoutElement(expr, el, ln)
                else:
                    raise ParseError("expected 'return' or 'element'", self.cur)

            elif self.at_kw("with"):
                self.kw("with")
                if self.at_kw("noise"):
                    self.kw("noise")
                    self.kw("across")
                    s1 = self.ident()
                    self.punct(",")
                    s2 = self.ident()
                    self.punct("(")
                    amp = self.number()
                    self.punct(")")
                    expr = WithNoise(expr, s1, s2, amp, ln)
                else:
                    el = self.ident()
                    self.kw("scaling")
                    expr = WithScaling(expr, el, self.number(), ln)

            else:  # reroute
                self.kw("reroute")
                self.kw("return")
                self.punct("(")
                frm = self.ident()
                self.punct(")")
                self.kw("through")
                expr = Reroute(expr, frm, self.parse_path(), ln)

        return expr


def parse(src: str) -> Program:
    return Parser(tokenize(src)).parse_program()

File: musculo-skeletal/web/src/lang/parser.ts
/**
 * Recursive-descent parser. The grammar is LL(1), so one token of
 * lookahead suffices and no backtracking is used.
 *
 * Both `outbound` and `return` are mandatory in a circuit declaration, so
 * an open circuit cannot arise by omission -- only through an explicit
 * `without` / `reroute` operator inside an experiment.
 */

import {
  type AntagonistDecl, type BindDecl, type CircuitDecl, type CircuitExpr,
  type ElementDecl, type EventTypeDecl, type ExperimentDecl,
  type FloorSpec, type InstanceDecl, type LesionDecl, type Observable,
  type PhaseDecl, type Program, type Quantity, type Span,
  type TemplateDecl, emptyProgram,
} from "./ast";
import { STRATA, type Stratum, type Token, tokenize } from "./lexer";

export class ParseError extends Error {
  constructor(msg: string, public token: Token) {
    super(`${msg} (got '${token.text || "<eof>"}' at line ${token.line}, col ${token.col})`);
  }
}

class Parser {
  private i = 0;
  private toks: Token[];

  constructor(toks: Token[]) {
    // Comments never reach the parser.
    this.toks = toks.filter((t) => t.kind !== "comment");
  }

  private get cur(): Token {
    return this.toks[this.i];
  }

  private span(t: Token = this.cur): Span {
    return { line: t.line, col: t.col, start: t.start, end: t.end };
  }

  private at(kind: Token["kind"], text?: string): boolean {
    const t = this.cur;
    return t.kind === kind && (text === undefined || t.text === text);
  }

  private atKw(...words: string[]): boolean {
    return this.cur.kind === "keyword" && words.includes(this.cur.text);
  }

  private eat(kind: Token["kind"], text?: string): Token {
    if (!this.at(kind, text)) {
      throw new ParseError(`expected ${text ?? kind}`, this.cur);
    }
    return this.toks[this.i++];
  }

  private kw(word: string) {
    return this.eat("keyword", word);
  }
  private punct(ch: string) {
    return this.eat("punct", ch);
  }

  /** Stratum names are keywords but legal in identifier position. */
  private ident(): string {
    if (this.cur.kind === "keyword" && (STRATA as readonly string[]).includes(this.cur.text)) {
      return this.toks[this.i++].text;
    }
    return this.eat("ident").text;
  }

  /** Joint names carry dots and digits (`muscle_Bicep.30_30`), so a binding
   *  target may be written as a quoted string. A bare identifier is also
   *  accepted for the rigs whose names happen to be identifier-shaped. */
  private stringOrIdent(): string {
    if (this.cur.kind === "string") {
      const t = this.toks[this.i++];
      return t.text.slice(1, -1);
    }
    return this.ident();
  }

  private quantity(): Quantity {
    const t = this.cur;
    if (t.kind === "quantity") {
      this.i++;
      return { value: t.value!, unit: t.unit! };
    }
    if (t.kind === "number") {
      this.i++;
      return { value: t.value!, unit: "" };
    }
    throw new ParseError("expected a quantity", t);
  }

  private number(): number {
    const t = this.cur;
    if (t.kind === "number" || t.kind === "quantity") {
      this.i++;
      return t.value!;
    }
    throw new ParseError("expected a number", t);
  }

  // ── program ───────────────────────────────────────────────────────

  parseProgram(): Program {
    const prog = emptyProgram();

    if (this.atKw("module")) {
      this.kw("module");
      prog.module = this.ident();
      this.punct(";");
    }
    while (this.atKw("import")) {
      this.kw("import");
      prog.imports.push(this.ident());
      this.punct(";");
    }

    while (!this.at("eof")) {
      if (this.atKw("compartment")) {
        prog.compartments.push(this.parseCompartment());
      } else if (this.atKw("circuit")) {
        const d = this.parseCircuitLike();
        if (d.kind === "template") prog.templates.push(d);
        else if (d.kind === "instance") prog.instances.push(d);
        else prog.circuits.push(d);
      } else if (this.atKw("event")) {
        prog.eventTypes.push(this.parseEventType());
      } else if (this.atKw("antagonist")) {
        prog.antagonists.push(this.parseAntagonist());
      } else if (this.atKw("bind")) {
        prog.binds.push(this.parseBind());
      } else if (this.atKw("experiment")) {
        prog.experiments.push(this.parseExperiment());
      } else {
        throw new ParseError("expected a declaration", this.cur);
      }
    }
    return prog;
  }

  private parseCompartment() {
    const t0 = this.kw("compartment");
    const name = this.ident();
    this.punct("{");
    this.kw("capacitance");
    this.punct(":");
    const capacitance = this.quantity();
    this.punct(";");
    this.kw("stratum");
    this.punct(":");
    if (!this.atKw(...STRATA)) throw new ParseError("expected a stratum name", this.cur);
    const stratum = this.toks[this.i++].text as Stratum;
    this.punct(";");
    this.punct("}");
    return { kind: "compartment" as const, name, capacitance, stratum, span: this.span(t0) };
  }

  private parseCircuitLike(): CircuitDecl | TemplateDecl | InstanceDecl {
    const t0 = this.kw("circuit");

    if (this.atKw("template")) {
      this.kw("template");
      const name = this.ident();
      this.punct("(");
      const params: string[] = [];
      if (!this.at("punct", ")")) {
        params.push(this.ident());
        while (this.at("punct", ",")) {
          this.punct(",");
          params.push(this.ident());
        }
      }
      this.punct(")");
      const body = this.parseCircuitBody();
      return { kind: "template", name, params, ...body, span: this.span(t0) };
    }

    const name = this.ident();

    if (this.at("punct", "=")) {
      this.punct("=");
      const template = this.ident();
      this.punct("(");
      const args: (string | Quantity)[] = [];
      if (!this.at("punct", ")")) {
        args.push(this.templateArg());
        while (this.at("punct", ",")) {
          this.punct(",");
          args.push(this.templateArg());
        }
      }
      this.punct(")");
      this.punct(";");
      return { kind: "instance", name, template, args, span: this.span(t0) };
    }

    const body = this.parseCircuitBody();
    return { kind: "circuit", name, ...body, span: this.span(t0) };
  }

  private templateArg(): string | Quantity {
    if (this.cur.kind === "number" || this.cur.kind === "quantity") return this.quantity();
    return this.ident();
  }

  private parseCircuitBody() {
    this.punct("{");
    this.kw("floor");
    this.punct(":");
    const floor = this.parseFloorSpec();
    this.punct(";");
    this.kw("outbound");
    this.punct(":");
    const outbound = this.parsePath();
    this.punct(";");
    this.kw("return");
    this.punct(":");
    const ret = this.parsePath();
    this.punct(";");

    const elements: ElementDecl[] = [];
    while (this.atKw("element")) elements.push(this.parseElement());
    this.punct("}");
    return { floor, outbound, ret, elements };
  }

  private parseFloorSpec(): FloorSpec {
    if (this.atKw("derived")) {
      this.kw("derived");
      this.punct("(");
      const derivedCall = this.ident();
      let derivedArg: string | undefined;
      if (this.at("punct", "(")) {
        this.punct("(");
        derivedArg = this.ident();
        this.punct(")");
      }
      this.punct(")");
      return { derivedCall, derivedArg };
    }
    return { literal: this.quantity() };
  }

  private parsePath(): string[] {
    const path = [this.ident()];
    while (this.at("arrow")) {
      this.eat("arrow");
      path.push(this.ident());
    }
    return path;
  }

  private parseElement(): ElementDecl {
    const t0 = this.kw("element");
    const name = this.ident();
    this.kw("conducts");
    const src = this.ident();
    this.eat("arrow");
    const dst = this.ident();

    let delay: Quantity | undefined;
    let gain = 1.0;
    if (this.atKw("delay")) {
      this.kw("delay");
      delay = this.quantity();
    }
    if (this.atKw("gain")) {
      this.kw("gain");
      gain = this.number();
    }
    this.punct(";");
    return { name, src, dst, delay, gain, span: this.span(t0) };
  }

  private parseEventType(): EventTypeDecl {
    const t0 = this.kw("event");
    this.kw("type");
    const name = this.ident();
    this.punct("=");
    const ctor = this.ident();
    const args: string[] = [];
    if (this.at("punct", "(")) {
      this.punct("(");
      if (!this.at("punct", ")")) {
        args.push(this.ident());
        while (this.at("punct", ",")) {
          this.punct(",");
          args.push(this.ident());
        }
      }
      this.punct(")");
    }
    this.punct(";");
    return { kind: "eventType", name, ctor, args, span: this.span(t0) };
  }

  private parseAntagonist(): AntagonistDecl {
    const t0 = this.kw("antagonist");
    const name = this.ident();
    this.punct("{");
    this.kw("agonist");
    this.punct(":");
    const agonist = this.ident();
    this.punct(";");
    this.kw("antagonist");
    this.punct(":");
    const antagonist = this.ident();
    this.punct(";");
    this.kw("shared");
    this.punct(":");
    const shared = [this.ident()];
    while (this.at("punct", ",")) {
      this.punct(",");
      shared.push(this.ident());
    }
    this.punct(";");
    this.punct("}");
    return { kind: "antagonist", name, agonist, antagonist, shared, span: this.span(t0) };
  }

  /**
   * bind <circuit> to rig("<name>") { comp -> joint; ... velocity: 50 m/s; }
   *
   * The binding lives in the source rather than in tool configuration
   * because it is checkable: the rig contributes an adjacency and a tissue
   * index the program does not have, so a binding can contradict the
   * circulation it binds. A contradiction is only reportable if the claim
   * was made in the language.
   */
  private parseBind(): BindDecl {
    const t0 = this.kw("bind");
    const circuit = this.ident();
    this.kw("to");
    this.kw("rig");
    this.punct("(");
    const rig = this.stringOrIdent();
    this.punct(")");
    this.punct("{");

    const map: Record<string, string> = {};
    let conductionVelocity: number | undefined;
    let unitsPerMetre: number | undefined;

    while (!this.at("punct", "}")) {
      if (this.atKw("velocity")) {
        this.kw("velocity");
        this.punct(":");
        conductionVelocity = this.quantity().value;
        this.punct(";");
        continue;
      }
      if (this.atKw("units")) {
        this.kw("units");
        this.punct(":");
        unitsPerMetre = this.quantity().value;
        this.punct(";");
        continue;
      }
      const comp = this.ident();
      this.eat("arrow");
      const joint = this.stringOrIdent();
      this.punct(";");
      map[comp] = joint;
    }
    this.punct("}");
    return { kind: "bind", circuit, rig, map, conductionVelocity, unitsPerMetre, span: this.span(t0) };
  }

  private parseExperiment(): ExperimentDecl {
    const t0 = this.kw("experiment");
    const name = this.ident();
    this.punct("{");
    this.kw("intact");
    this.punct(":");
    const intact = this.parseCircuitExpr();
    this.punct(";");

    const lesions: LesionDecl[] = [];
    let observables: Observable[] = [];
    const phases: PhaseDecl[] = [];

    while (!this.at("punct", "}")) {
      if (this.atKw("lesion")) lesions.push(this.parseLesion());
      else if (this.atKw("phase")) phases.push(this.parsePhase());
      else if (this.atKw("observe")) {
        this.kw("observe");
        this.punct(":");
        observables = this.parseObservables();
        this.punct(";");
      } else throw new ParseError("expected lesion, phase, or observe", this.cur);
    }
    this.punct("}");

    if (phases.length && (lesions.length || observables.length)) {
      throw new ParseError("an experiment is either phased or flat, not both", this.cur);
    }
    return { kind: "experiment", name, intact, lesions, observables, phases, span: this.span(t0) };
  }

  private parsePhase(): PhaseDecl {
    const t0 = this.kw("phase");
    const name = this.ident();
    let fromPhase: string | undefined;
    if (this.atKw("from")) {
      this.kw("from");
      fromPhase = this.ident();
    }
    this.punct("{");
    const lesions: LesionDecl[] = [];
    let observables: Observable[] = [];
    while (!this.at("punct", "}")) {
      if (this.atKw("lesion")) lesions.push(this.parseLesion());
      else if (this.atKw("observe")) {
        this.kw("observe");
        this.punct(":");
        observables = this.parseObservables();
        this.punct(";");
      } else throw new ParseError("expected lesion or observe", this.cur);
    }
    this.punct("}");
    return { name, lesions, observables, fromPhase, span: this.span(t0) };
  }

  private parseLesion(): LesionDecl {
    const t0 = this.kw("lesion");
    const name = this.ident();
    this.punct(":");
    const expr = this.parseCircuitExpr();
    this.punct(";");
    return { name, expr, span: this.span(t0) };
  }

  private parseObservables(): Observable[] {
    const obs = [this.parseObservable()];
    while (this.at("punct", ",")) {
      this.punct(",");
      obs.push(this.parseObservable());
    }
    return obs;
  }

  private parseObservable(): Observable {
    const t0 = this.cur;
    const name = this.ident();
    const args: string[] = [];
    if (this.at("punct", "(")) {
      this.punct("(");
      if (!this.at("punct", ")")) {
        args.push(this.ident());
        while (this.at("punct", ",")) {
          this.punct(",");
          args.push(this.ident());
        }
      }
      this.punct(")");
    }
    return { name, args, span: this.span(t0) };
  }

  /** Left-associative chain of postfix lesion operators. */
  private parseCircuitExpr(): CircuitExpr {
    const t0 = this.cur;
    const sp = this.span(t0);
    let expr: CircuitExpr = { op: "ref", name: this.ident(), span: sp };

    while (this.atKw("without", "with", "reroute")) {
      const opTok = this.cur;
      const osp = this.span(opTok);

      if (this.atKw("without")) {
        this.kw("without");
        if (this.atKw("return")) {
          this.kw("return");
          this.punct("(");
          const from = this.ident();
          this.punct(")");
          expr = { op: "withoutReturn", base: expr, from, span: osp };
        } else if (this.atKw("element")) {
          this.kw("element");
          this.punct("(");
          const element = this.ident();
          this.punct(")");
          expr = { op: "withoutElement", base: expr, element, span: osp };
        } else throw new ParseError("expected 'return' or 'element'", this.cur);
      } else if (this.atKw("with")) {
        this.kw("with");
        if (this.atKw("noise")) {
          this.kw("noise");
          this.kw("across");
          const s1 = this.ident() as Stratum;
          this.punct(",");
          const s2 = this.ident() as Stratum;
          this.punct("(");
          const amplitude = this.number();
          this.punct(")");
          expr = { op: "withNoise", base: expr, s1, s2, amplitude, span: osp };
        } else {
          const element = this.ident();
          this.kw("scaling");
          expr = { op: "withScaling", base: expr, element, factor: this.number(), span: osp };
        }
      } else {
        this.kw("reroute");
        this.kw("return");
        this.punct("(");
        const from = this.ident();
        this.punct(")");
        this.kw("through");
        expr = { op: "reroute", base: expr, from, path: this.parsePath(), span: osp };
      }
    }
    return expr;
  }
}

export function parse(src: string): Program {
  return new Parser(tokenize(src)).parseProgram();
}
